Feature: Matching interface with name

cathay4t opened the pull request:
Introducing two properties to base interface:
 * `identifier` -- `name` or `mac-address`, default to `name`
 * `profile-name`: string

When identifier been set to `mac-address`, nmstate will not use
`interface.name` for NIC matching but use the `interface.mac-address`, The
`interface.name` will be used for `profile-name` when storing the network
configuration in backend.

The `profile-name` will be hide if it is equal to interface name.

Example on creating MAC based profile:

```yml
---
interfaces:
- name: wan0
  type: ethernet
  state: up
  identifier: mac-address
  mac-address: 1e:bd:23:e9:fb:94
```

You may delete above profile using `name: wan0` or its real interface name.

Integration test cases included.

cathay4t committed:
Run cargo fmt, cargo clippy and python black

Signed-off-by: Gris Ge <fge@redhat.com>

jhernand commented:
In case it helps I wanted to share that I tested this change and it works fine. I created a configuration file like this:

```yaml
interfaces:
- name: my
  identifier: mac-address
  mac-address: 02:00:00:00:00:02
  type: ethernet
  state: up
  ipv4:
    enabled: true
    address:
    - ip: 192.168.100.10
      prefix-length: 24
  ipv6:
    enabled: false
dns-resolver:
  config:
    search:
    - home.arpa
    server:
    - 192.168.100.1
```

And it generates the following NetworkManager connection file:

```ini
[connection]
autoconnect=true
autoconnect-slaves=-1
id=my
type=802-3-ethernet
uuid=e75f0ca9-1672-56f0-990d-51f87614a2f6

[ipv4]
address0=192.168.100.10/24
dns=192.168.100.1
dns-priority=40
dns-search=home.arpa
method=manual

[ipv6]
addr-gen-mode=0
method=disabled

[ethernet]
mac-address=02:00:00:00:00:02
```

That means that the NetworkManager connection file no longer depends on the name assigned by the kernel/udev/systemd.

Thanks @cathay4t!

cathay4t commented:
~~Investigating CI failure~~

CI problem resolved. It was caused by a leftover of VLAN test case.

cathay4t committed:
test: Remove leftover of `test_add_qinq_vlan`

The test case `test_add_qinq_vlan` forgot to remove the interface on
finish.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
Feature: Matching interface with name

Introducing two properties to base interface:
 * `identifier` -- `name` or `mac-address`, default to `name`
 * `profile-name`: string

When identifier been set to `mac-address`, nmstate will not use
`interface.name` for NIC matching but use the `interface.mac-address`,
The `interface.name` will be used for `profile-name` when storing the
network configuration in backend.

The `profile-name` will be hide if it is equal to interface name.

Example on creating MAC based profile:

```yml
---
interfaces:
  - name: wan0
    type: ethernet
    state: up
    identifier: mac-address
    mac-address: 1e:bd:23:e9:fb:94
```

You may delete above profile using `name: wan0` or its real interface
name.

Integration test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>
